doc/designs/prov_loadex.md

```diff
@@ -0,0 +1,78 @@
+Providers run-time configuration
+================================
+
+Currently any provider run-time activation requires presence of the
+initialization parameters in the OpenSSL configuration file. Otherwise the
+provider will be activated with some "default" settings, that may or may not
+work for a particular application. For real-world systems it may require
+providing a specially designed OpenSSL config and passing it somehow (e.g. via
+environment) that has its obvious drawbacks.
+
+We need a possibility to initialize providers on per-application level
+according to per-application parameters. It's necessary for example for PKCS#11
+provider (where different applications may use different devices with different
+drivers) and will be useful for some other providers. In case of Red Hat it is
+also usable for FIPS provider.
+
+OpenSSL 3.2 introduces the API
+
+```C
+OSSL_PROVIDER *OSSL_PROVIDER_load_ex(OSSL_LIB_CTX *libctx, const char *name,
+                                     OSSL_PARAM params[]);
+```
+
+intended to configure the provider in load time.
+
+It accepts only parameters of type `OSSL_PARAM_UTF8_STRING` because any
+provider can be initialized from the config file where the values are
+represented as strings and provider init function has to deal with it.
```

Comment:
The values from a configuration file are necessarily UTF8 strings because of its origin.  I'm not sure that's a reason why we *have* to limit parameters passed this way to UTF8 strings, though.  There is, of course, the caveat that non-UTF8 strings will not be possible to replicate into a config file...  question is, is that terrible?

Reply:
I don't expect any problems while we have null-terminated strings here :)

Reply:
I frankly don't see it as a *restriction*, but fair enough.  I think it is important to have some level of expectation on humanly readable strings, and UTF8 was a choice that I based on how universal that encoding of non-ASCII chars has become.

Reply:
Exactly. The current internal representation is C string (and UTF8 works as C string) so it just a simple solution.